search facets + historyservice + server-side graph save

kfoss opened the pull request.

Commit:
Update input-select style

Commit:
Add wait-spinner icon for async actions

Commit:
Update histogram to use Angular POST syntax

erickt commented on src/main/java/org/lab41/dendrite/services/HistoryService.java:
```diff
@@ -0,0 +1,41 @@
+        file.mkdirs();
+    }
+
+    public Configuration configuration;
```
I don't think you need to keep a hold of the configuration object. Instead can you do:

```
private String historyStorage;

public String getHistoryStorage {
    return historyStorage;
}

public HistoryService(...) {
    ...
    historyStorage = configuration.getString("history.storage");
    File file = new File(historyStorage);
    file.mkdirs();
}
```

erickt commented on src/main/java/org/lab41/dendrite/web/controller/GraphExportController.java:
```diff
@@ -83,4 +88,48 @@ public void export(@PathVariable String graphName, GraphExportBean exportItem, H
+            return;
+        }
+
+        Graph graph = application.getGraph(graphName);
```
It turns out titan doesn't automatically rollback or commit the implicit transaction that's created here. I'll mark the places where this needs to be done.

erickt commented on src/main/java/org/lab41/dendrite/web/controller/GraphExportController.java:
```diff
@@ -83,4 +88,48 @@ public void export(@PathVariable String graphName, GraphExportBean exportItem, H
+            }
+        } catch (IOException e) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        }
```
This needs a `graph.commit()`.

erickt commented on src/main/java/org/lab41/dendrite/web/controller/viz/ElasticSearchController.java:
```diff
@@ -33,6 +50,115 @@
+
+        if (!(graph instanceof TitanGraph)) {
+            json.put("status", "error");
+            json.put("error", "graph is not a titan graph");
```
This needs a `graph.rollback()`

erickt commented on src/main/java/org/lab41/dendrite/web/controller/viz/ElasticSearchController.java:
```diff
@@ -33,6 +50,115 @@
+
+        if (indexName == null) {
+            json.put("status", "error");
+            json.put("msg", "unknown index '" + indexName + "'");
```
This needs a `graph.rollback()`

erickt commented on src/main/java/org/lab41/dendrite/web/controller/viz/ElasticSearchController.java:
```diff
@@ -33,6 +50,115 @@
+        // if it didn't find either index hostname or local-mode for the specified index
+        if (flag == 0) {
+            json.put("status", "error");
+            json.put("msg", "ElasticSearch index not found '" + indexName + "'");
```
This needs a `graph.rollback()`

erickt commented on src/main/java/org/lab41/dendrite/web/controller/viz/ElasticSearchController.java:
```diff
@@ -33,6 +50,115 @@
+            jsonResult = new JSONObject(content);
+        }catch (Exception ex) {
+            json.put("status", "error");
+            json.put("msg", "error retrieving query from ElasticSearch");
```
This needs a `graph.rollback()`

erickt commented on src/main/java/org/lab41/dendrite/web/controller/viz/ElasticSearchController.java:
```diff
@@ -33,6 +50,115 @@
+        } finally {
+            httpClient.getConnectionManager().shutdown();
+        }
+
```
This needs a `graph.commit()`

Commit:
Convert search facets to select dropdown

Commit:
Add shell HistoryService

Commit:
Add backend class to save graph server-side

API endpoint "file-save" saves the graph to the
HistoryService-defined storage location, which is
envisioned to be a .git repository for committing
changes/actions captured through the UI.

Commit:
Add license headers

Commit:
Add UI component to save graph server-side
